Fixing files for deployment of ASLOv3

vipulgupta2048 opened the pull request:
Signed-off-by: Vipul Gupta (@vipulgupta2048) <vipulgupta2048@gmail.com>

In reference to the project in GSoC 2018,
For the successful deployment of ASLOv3, some files and activity.info need to be added/fixed/updated to present correct metadata on the website.

vipulgupta2048 committed:
Update metadata -  activity.info

Added website, categories data to the activity.info

Signed-off-by: Vipul Gupta (@vipulgupta2048) <vipulgupta2048@gmail.com>

vipulgupta2048 commented:
@quozl This would be the first PR of this new project. Let's form a new template for PR's and Commits that I could follow in the future and make our work a little easier. Hence change or request changes in anything you would want in this PR. Thanks !!

walterbender commented:
I am a bit confused by the addition of a LICENSE file when we already have COPYING.

vipulgupta2048 commented:
The presence of License file came up false in the script, hence I added it up without thinking much.
Also, repositories do need a separate License file, hence I added it up because of that. 
Should I remove it? 
or I can just add another parameter to the script that could look for COPYING too

walterbender commented:
I think the script needs to check for either LICENSE or COPYING. We don't need both.

vipulgupta2048 commented:
COPYING it is then, removing LICENSE file from the PR.
But what if they are both in a repo? #justsaying

vipulgupta2048 committed:
Changes made as requested

Signed-off-by: Vipul Gupta (@vipulgupta2048) <vipulgupta2048@gmail.com>

walterbender commented:
There should not be both, but it there are, keep the one that was added by the activity author, which will likely be COPYING, as that has been our convention.

vipulgupta2048 commented:
Understood, any comments on the activity.info file changes?

walterbender commented:
That looks fine to me.

quozl commented on LICENSE:
```diff
@@ -0,0 +1,675 @@
+### GNU GENERAL PUBLIC LICENSE
```
Do not add this file.

* `COPYING` is the most correct name to use,
* `COPYING` already is in the repository and is consistent with the source files license declaration, so `LICENSE` would be duplication,
* `activity/activity.info` also has correct license name.

quozl commented on README.md:
```diff
@@ -6,7 +6,7 @@ Abacus is a tool for simple arithmetic calculations.
 ===========
 
-Abacus is part of the Sugar desktop.  Please refer to;
+Abacus is part of the Sugar desktop. Please refer to;
```
Change is not needed.  Two spaces after a period are part of some writing style guides, and they are frequently used and harmless.

vipulgupta2048 replied:
Implemented.

quozl replied:
Change is not visible, please push.

quozl commented on README.md:
```diff
@@ -31,3 +31,8 @@ dnf install sugar-abacus
 Abacus is started by [Sugar](https://github.com/sugarlabs/sugar).
+
+License
+=======
```
Do not add license to README.md.  It is duplicate information that is not necessary.  README.md is always presented next to a file list by GItHub, so COPYING is quite visible.

vipulgupta2048 replied:
Removed.

quozl replied:
Change is not visible, please push.

quozl commented on activity/activity.info:
```diff
@@ -6,6 +6,7 @@ bundle_id = org.sugarlabs.AbacusActivity
-category = maths
-repository = https://github.com/sugarlabs/abacus
\ No newline at end of file
+website = https://help.sugarlabs.org/en/abacus.html
```
`website` is not defined in the metadata reference.  See https://github.com/sugarlabs/sugar-toolkit-gtk3/blob/master/src/sugar3/bundle/__init__.py and the older https://wiki.sugarlabs.org/go/Development_Team/Almanac/Activity_Bundles#.info_file_format

Before adding `website`, I'd like a patch to `bundle/__init__.py` to add the definition.

vipulgupta2048 replied:
I was told to reference this https://wiki.sugarlabs.org/go/Development_Team/Almanac/Activity_Bundles#.info_file_format but I was also referring to the active Turtleart activity which had this https://github.com/sugarlabs/turtleart-activity/blob/master/activity/activity.info

vipulgupta2048 replied:
`Website` could be the placeholder for the help-activity links as we have previously discussed. Hence I suggest we add it to `bundle/__init__.py` and also edit the wiki

quozl replied:
Thanks.  I'm not surprised you were pointed at only one of the lists of metadata names, but it is only one of the lists; the list in the toolkit is just as important.  Being told is no excuse.  :grin:  Please feed this back to whoever _told_ you.  I'd like to see the Wiki page deprecated in favour of the toolkit documentation.

Yes, I am aware some activities have begun to use `website` before adding it to the lists.  The activities are Turtle Blocks and derivatives, as well as Music Blocks.  I'm sure Walter would be willing to propose the change too, but as you're the one working on this wider issue, you are best placed for this.

While you are at it, you might also;

* make changes for other, as yet "unofficial", names such as `categories`,
* resolve any differences between the Wiki and the toolkit, and;
* deprecate the Wiki page.

quozl replied:
Reviewed again; waiting for patch to `bundle/__init__.py` in toolkit.

quozl commented on activity/activity.info:
```diff
@@ -6,6 +6,7 @@ bundle_id = org.sugarlabs.AbacusActivity
\ No newline at end of file
+website = https://help.sugarlabs.org/en/abacus.html
+summary = A tool for simple arithmetic calculations
+categories = maths
```
`categories` is not defined in the metadata reference, please patch `bundle/__init__.py` to add it as an alias for `tags`, and allow either semicolon <kbd>;</kbd> or whitespace delimiters.

Please also add a list of categories to choose from, or a pointer in `bundle/__init__.py` to where the categories are listed in another repository.

In a sample of 78 working activities, 19 use `category` and 8 use `categories`.  None of them used `tags`

vipulgupta2048 replied:
Before patching the `bundle/__init__.py` 
We can either define right here on what to use and I can make them the same everywhere.  `Categories` or `Category` irrespective of the grammatical part

quozl replied:
Sure, if you like, but you'll only be defining it with Walter and I.  Once you make a pull request for `bundle/__init__.py` you may get different reviewers.  We have developers who watch the toolkit repository more closely than this considerably isolated activity.  20 vs 6 watching.

My vote is for the name to be any of these three names `tags`, `category`, `categories`, and for the delimiter to be either semicolon or whitespace.  Since nothing in the toolkit uses the name, it is of little consequence except as standardisation.

quozl reviewed (changes requested):
Thanks for the patch.  Please make changes as requested, and also;

* improve the commit message instead of just "Updated activity.info"

quozl commented:
> But what if they are both in a repo?

This does happen, and it may not be a problem.

Please do not write code that depends on or processes either LICENSE or COPYING.  The purpose of these files is _not to define the licenses_, but to _supply a copy of the licenses_ as part of how a developer applies a license to their work.

The copyright and license of source files are in the source files.

The `license` metadata in `activity/activity.info` summarises the licenses of the source files in the repository; as a semicolon <kbd>;</kbd> separated list.  Depending on the licenses used, there may be files such as `COPYING` (for GPLv2, GPLv2+, GPLv3, GPLv3+), or `COPYING.LGPL` (for LGPLv2), or any number of files that _supply a copy of a license_.  These files can also be in subdirectories, and not the main directory of a repository.

@vipulgupta2048, please, as part of your research, fully read the GPLv3 text from top to bottom, noting especially how a license is _applied_ by the originating developer.

There are software tools for reconciling copyrights and licenses; these are used by Debian and Fedora projects to automate packaging.  See the Ubuntu packages `licensecheck` and `license-reconcile`.  You should be able to find similar libraries in whatever scripting language you are using for your work.

vipulgupta2048 commented:
Thanks for the review, I will implement the changes requested and read/research the material you referred to, then get back to you.

quozl commented:
@vipulgupta2048 wrote:
> For the successful deployment of ASLOv3, some files and activity.info need to be added/fixed/updated to present correct metadata on the website.

Please do not stall your project on updates to metadata in activities; at one update a week it would take your whole project time period and only a few activities would be available.

Instead, please design or change ASLOv3 to assume useful default values and not show missing metadata.

For link to the Help Activity, you might read the `bundle_id` instead and find it in the `helplink.json` file in the Help repository.

For link to repository, you might read the `bundle_id`, find it in GitHub, and provide a link that way.

vipulgupta2048 commented:
> Please do not write code that depends on or processes either LICENSE or COPYING. The purpose of these files is not to define the licenses, but to supply a copy of the licenses as part of how a developer applies a license to their work.

My script isn't depending on the files mentioned above. But just making sure that they are present in the repository and that the contents of COPYING match the license field in activity.info as Walter stated.

quozl commented:
> My script isn't depending on the files mentioned above. But just making sure that they are present in the repository and that the contents of COPYING match the license field in activity.info as Walter stated.

Oh, okay.  Sorry, I was relying on what you first said; that successful deployment of ASLOv3 needs these changes.  Now I know you didn't mean that.

There should indeed be copies of the license in a repository; and there should indeed be a `license` metadata field, but stalling ASLOv3 for these things doesn't seem scalable.

vipulgupta2048 commented:
To be honest, there was a lot to take in with your review, hence after proper thinking, I am going forward with the following. 
1. ~Make a PR for patch to `bundle/__init__.py` in toolkit.~
2. ~Push my local changes that you requested.~
3. ~Let's discuss the depreciation of the aforementioned wiki page.~ 
4. ~Did read what you suggested and checked out `licensecheck` of Debian. I have been inactive for somedays, due to personal problem. No matter what, I wouldn't be stalling the project.~ 
Following tasks has been completed with the PR updated as changes requested.
Thanks !!

vipulgupta2048 commented:
Changes made as requested, also changed previous commit message.  
- Taking back from this PR, I will make the fixing and updating of files my top priority and then later on work through ASLOv3 to show default values directly from help-activity if not present as you suggested.

vipulgupta2048 commented:
@quozl Any other changes that are needed?

vipulgupta2048 commented:
Thanks @quozl

vipulgupta2048 commented:
> Instead, please design or change ASLOv3 to assume useful default values and not show missing metadata.

Regarding this, I contacted Jatin who worked on ASLO previously and get some insights on the code he has written. 

> There are software tools for reconciling copyrights and licenses; these are used by Debian and Fedora projects to automate packaging. See the Ubuntu packages licensecheck and license-reconcile. You should be able to find similar libraries in whatever scripting language you are using for your work.

Regarding this, very interesting stuff. But I couldn't find one that could be used for Python Scripts. Can you point me towards one, if you have used it. Thanks.

quozl commented:
> > Instead, please design or change ASLOv3 to assume useful default values and not show missing metadata.

> Regarding this, I contacted Jatin who worked on ASLO previously and get some insights on the code he has written. 

What Jatin wrote should be changed; that's easy isn't it?  Look for where to change it.

> > There are software tools for reconciling copyrights and licenses; these are used by Debian and Fedora projects to automate packaging. See the Ubuntu packages licensecheck and license-reconcile. You should be able to find similar libraries in whatever scripting language you are using for your work.

> Regarding this, very interesting stuff. But I couldn't find one that could be used for Python Scripts. Can you point me towards one, if you have used it. Thanks. 

What do you mean?  licensecheck can be used to inspect Python, and JavaScript, which are the two languages most frequently used by Sugar activities.  If you mean "how do I use them from within a Python script", there's `subprocess.check_output()` as a minimum, and various other APIs for calling Perl programs from within Python.  But `subprocess` is much easier.

Install and try `licensecheck` yourself.  Look at `licensecheck --help` for options.  The `--machine` option will help for processing output.  The `--recursive` option will help for processing all input files in a directory.
